File: c/parallel/gram_schmidt.c
#include "gram_schmidt.h"
#include <mpi.h>
#include <math.h>
#include <stdlib.h>

static double dot_product(double *A, double *B, int n, MPI_Comm comm)
{
    int k;
    double sum = 0.;

    #pragma omp parallel for reduction(+:sum)
    for (k = 0; k < n; k++) {
        sum += A[k] * B[k];
    }

    MPI_Allreduce(MPI_IN_PLACE, &sum, 1, MPI_DOUBLE, MPI_SUM, comm);

    return sum;
}

static void multiply_add(double *A, double c, double *B, int n)
{
    int k;

    #pragma omp parallel for
    for (k = 0; k < n; k++) {
        A[k] += c * B[k];
    }
}

static void multiply(double *A, double c, int n)
{
    int k;

    #pragma omp parallel for
    for (k = 0; k < n; k++) {
        A[k] *= c; 
    }
}

#define CEIL_DIV(x, y) (((x) + (y) - 1) / (y))

void gram_schmidt(double *vector[])
{
    int i, j, r;
    double coef;
    double *v[N];
    
    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    
    int unit = CEIL_DIV(M, size);
    int nprocs = CEIL_DIV(M, unit);

    MPI_Group world_group;
    MPI_Comm_group(MPI_COMM_WORLD, &world_group);
    MPI_Group work_group;
    int range[3] = {0, nprocs - 1, 1};
    MPI_Group_range_incl(world_group, 1, &range, &work_group);
    MPI_Comm comm;
    MPI_Comm_create(MPI_COMM_WORLD, work_group, &comm);
    MPI_Group_free(&world_group);
    MPI_Group_free(&work_group);
    
    if (rank >= nprocs) return;
    MPI_Comm_rank(comm, &rank);

    int counts[nprocs], displs[nprocs];

    for (r = 0; r < nprocs; r++) {
        if (r != nprocs - 1) {
            counts[r] = unit;
        } else {
            counts[r] = M - unit*r;
        }
        displs[r] = unit*r;
    }
    
    int chunk = counts[rank];

    for (i = 0; i < N; i++) {
        v[i] = malloc(sizeof(double) * chunk);
    }

    for (i = 0; i < N; i++) {
        MPI_Scatterv(vector[i], counts, displs, MPI_DOUBLE, v[i], chunk, MPI_DOUBLE, 0, comm);
    }

    for (j = 0; j < N; j++) {
        for (i = 0; i < j; i++) {
            coef = -dot_product(v[i], v[j], chunk, comm);
            multiply_add(v[j], coef, v[i], chunk);
        }
        coef = 1. / sqrt(dot_product(v[j], v[j], chunk, comm));
        multiply(v[j], coef, chunk);
    }

    for (i = 0; i < N; i++) {
        MPI_Gatherv(v[i], chunk, MPI_DOUBLE, vector[i], counts, displs, MPI_DOUBLE, 0, comm);
    }

    for (i = 0; i < N; i++) {
        free(v[i]);
    }
    
    MPI_Comm_free(&comm);
}

File: fortran/parallel/gram_schmidt.f90
module gram_schmidt
  use mpi
  use param

  implicit none

  private :: dot_product, multiply_add, multiply, ceil_div
  
contains
  
  real(8) function dot_product(A, B, n, comm)
    real(8), dimension(0:n-1) :: A, B
    integer :: n, comm
    integer :: ierr
    integer :: k
    real(8) :: sum

    sum = 0.0_8
    !$omp parallel do reduction(+:sum)
    do k = 0, n - 1
      sum = sum + A(k)*B(k)
    enddo
    !$omp end parallel do
    
    call mpi_allreduce(mpi_in_place, sum, 1, mpi_real8, mpi_sum, comm, ierr)
    
    dot_product = sum
  end function dot_product
  
  subroutine multiply_add(A, c, B, n)
    real(8), dimension(0:n-1) :: A, B
    real(8) :: c
    integer :: n
    integer :: k
    
    !$omp parallel do
    do k = 0, n - 1
      A(k) = A(k) + c*B(k)
    enddo
    !$omp end parallel do
  end subroutine multiply_add
  
  subroutine multiply(A, c, n)
    real(8), dimension(0:n-1) :: A
    real(8) :: c
    integer :: n
    integer :: k
    
    !$omp parallel do
    do k = 0, n - 1
      A(k) = A(k)*c
    enddo
    !$omp end parallel do
  end subroutine multiply
  
  integer function ceil_div(x, y)
    integer :: x, y

    ceil_div = ((x) + (y) - 1) / (y)
  end function ceil_div
  
  subroutine gram_schmidt_process(vector)
    real(8), dimension(0:M-1,0:N-1) :: vector
    real(8), dimension(:,:), allocatable :: v
    integer :: i, j
    real(8) :: coef
    integer :: ierr, rank, size, unit, nprocs, world_group, work_group, range(3,1), comm, r, chunk
    integer, dimension(:), allocatable :: counts, displs
    
    call mpi_comm_rank(mpi_comm_world, rank, ierr)
    call mpi_comm_size(mpi_comm_world, size, ierr)
    
    unit = ceil_div(M, size)
    nprocs = ceil_div(M, unit)
    
    call mpi_comm_group(mpi_comm_world, world_group, ierr)
    range(:,1) = (/ 0, nprocs - 1, 1 /)
    call mpi_group_range_incl(world_group, 1, range, work_group, ierr)
    call mpi_comm_create(mpi_comm_world, work_group, comm, ierr)
    call mpi_group_free(world_group, ierr)
    call mpi_group_free(work_group, ierr)
    
    if (rank >= nprocs) return
    call mpi_comm_rank(comm, rank, ierr)

    allocate(counts(0:nprocs-1))
    allocate(displs(0:nprocs-1))

    do r = 0, nprocs - 1
      if (r /= nprocs - 1) then
        counts(r) = unit
      else
        counts(r) = M - unit*r
      endif
      displs(r) = unit*r
    enddo
    
    chunk = counts(rank)
    
    allocate(v(0:chunk-1,0:N-1))

    do i = 0, N - 1
      call mpi_scatterv(vector(:,i), counts, displs, mpi_real8, v(:,i), chunk, mpi_real8, 0, comm, ierr)
    enddo
    
    do j = 0, N - 1
      do i = 0, j - 1
        coef = -dot_product(v(:,i), v(:,j), chunk, comm)
        call multiply_add(v(:,j), coef, v(:,i), chunk)
      enddo
      coef = 1.0_8 / sqrt(dot_product(v(:,j), v(:,j), chunk, comm))
      call multiply(v(:,j), coef, chunk)
    enddo

    do i = 0, N - 1
      call mpi_gatherv(v(:,i), chunk, mpi_real8, vector(:,i), counts, displs, mpi_real8, 0, comm, ierr)
    enddo
    
    deallocate(v)
    deallocate(counts)
    deallocate(displs)
    call mpi_comm_free(comm, ierr)
  end subroutine gram_schmidt_process
  
end module gram_schmidt
